src/Google.Api.Gax/Platform.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Net.Http;
+using System.Threading;
+using System.Net;
+
+namespace Google.Api.Gax
+{
+    public enum PlatformType
+    {
+        Unknown = 0,
+        Gce = 1,
+        Gae = 2,
+    }
+
+    public sealed class GcePlatformDetails
+    {
+        public GcePlatformDetails(string metadataJson)
+        {
+            MetadataJson = metadataJson;
+            var json = Newtonsoft.Json.Linq.JObject.Parse(metadataJson);
+            ProjectId = json["project"]["projectId"].ToString();
+            InstanceId = json["instance"]["id"].ToString();
+            ZoneName = json["instance"]["zone"].ToString();
+        }
+
+        public string MetadataJson { get; }
+        public string ProjectId { get; }
+        public string InstanceId { get; }
+        public string ZoneName { get; }
+    }
+
+    public sealed class GaePlatformDetails
+    {
+        public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeVersion)
+        {
+            ProjectId = gcloudProject;
+
+        }
+
+        public string ProjectId { get; }
+        public string ModuleId { get; }
+        public string VersionId { get; }
+    }
+
+    /// <summary>
+    /// Information about the current execution platform.
+    /// Supported execption platforms are Google App Engine (GAE) and Google Compute Engine (GCE).
+    /// </summary>
+    public sealed class Platform
+    {
+        private readonly static Lazy<Task<Platform>> s_instance = new Lazy<Task<Platform>>(LoadInstance);
+
+        /// <summary>
+        /// Asyncrhonously get execution platform information.
+        /// </summary>
+        /// <returns>A task containing the execution platform information.</returns>
+        public static Task<Platform> InstanceAsync() => s_instance.Value;
+
+        /// <summary>
+        /// Get execution platform information. This may block for up to one second.
+        /// </summary>
+        /// <returns>Execution platform information.</returns>
+        public static Platform Instance() => InstanceAsync().Result;
+
+        private static async Task<GcePlatformDetails> LoadGceDetails()
+        {
+            const string metadataUrl = "http://metadata.google.internal/computeMetadata/v1?recursive=true";
```

Comment:
You may want to look for the METADATA_EMULATOR_HOST environment variable and use that value if present instead of "metadata.google.internal"